framework/src/controller/channels/base_channel.ts

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright © 2019 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { EventCallback } from '../event';
+import { Action, ActionsDefinition, ActionsObject } from '../action';
+import { INTERNAL_EVENTS, eventWithModuleNameReg } from '../constants';
+
+export interface BaseChannelOptions {
+	readonly skipInternalEvents?: boolean;
+}
+
+export abstract class BaseChannel {
+	public readonly moduleAlias: string;
+	public readonly options: object;
+
+	public readonly eventsList: ReadonlyArray<string>;
+	public readonly actionsList: ReadonlyArray<string>;
+	public readonly actions: ActionsObject;
+
+	public constructor(
+		moduleAlias: string,
+		events: ReadonlyArray<string>,
+		actions: ActionsDefinition,
+		options: BaseChannelOptions = {},
+	) {
+		this.moduleAlias = moduleAlias;
+		this.options = options;
+
+		const eventList = options.skipInternalEvents
+			? events
+			: [...events, ...INTERNAL_EVENTS];
+
+		this.eventsList = eventList;
+
+		this.actions = {};
+		for (const actionName of Object.keys(actions)) {
+			const actionData = actions[actionName];
+
+			const handler =
+				typeof actionData === 'object' ? actionData.handler : actionData;
+			const isPublic =
+				typeof actionData === 'object' ? actionData.isPublic ?? true : true;
+
+			const action = new Action(
+				`${this.moduleAlias}:${actionName}`,
+				undefined,
+				undefined,
+				isPublic,
+				handler,
+			);
+			this.actions[actionName] = action;
+		}
+		this.actionsList = Object.keys(this.actions);
+	}
+
+	public isValidEventName(name: string, throwError = true): boolean | never {
```

Comment:
why does this have `never` type?

Reply:
It throws an error if the event name is not valid.

Reply:
I think throwing is not consider as return type

Reply:
Throwing is not return type but `never` is. It is used to specify that function could throw an error in that case it will not return anything. https://www.typescriptlang.org/docs/handbook/basic-types.html#never